building/glass/upload.py

```diff
@@ -123,20 +119,43 @@ def gs_rsync(local_path: str, remote_path: str, boto_path: str):
 """.lstrip()
 
 
-def perform_uploads(uploads: dict, branch: str) -> None:
+def upload_gs(staging, root: str, branch_config: aptbranch.Config):
+    branch = branch_config.name
+    if not GS_BRANCH_REGEX.match(branch):
+        raise Exception("not an uploadable branch: %s" % branch)
     if not os.path.exists(BOTO_PATH):
         raise Exception("you need to put the GCP service account private key file into /homeworld/boto-key")
+    botoconfig_name = os.path.join(staging, "boto.config")
+    with open(botoconfig_name, "w") as bout:
+        with open(BOTO_PATH, "r") as f:
+            project_id = json.load(f)["project_id"]
+        bout.write(BOTO_TEMPLATE % (BOTO_PATH, project_id))
+        bout.flush()
+    gs_rsync(root, branch, botoconfig_name)
+
+
+def upload_rsync(staging, root: str, branch_config: aptbranch.Config):
+    target_dir = os.path.join(branch_config.upload_config["dir"], branch_config.name)
+    if ":" in branch_config.upload_config["target"]:
+        raise Exception("colon in rsync upload target")
+    dest = "%s:%s" % (branch_config.upload_config["target"], target_dir)
+    subprocess.check_call(["rsync", "-av", "--progress", "--delete-delay", "--", root + "/", dest])
+
+
+def perform_uploads(uploads: dict, branch_config: aptbranch.Config) -> None:
     with tempfile.TemporaryDirectory() as staging:
         root = os.path.join(staging, "root")
         for remote_path, local_path in uploads.items():
             target = os.path.join(root, remote_path.lstrip("/"))
             if not os.path.isdir(os.path.dirname(target)):
                 os.makedirs(os.path.dirname(target))
             shutil.copy2(local_path, target)
-        botoconfig_name = os.path.join(staging, "boto.config")
-        with open(botoconfig_name, "w") as bout:
-            with open(BOTO_PATH, "r") as f:
-                project_id = json.load(f)["project_id"]
-            bout.write(BOTO_TEMPLATE % (BOTO_PATH, project_id))
-            bout.flush()
-        gs_rsync(root, branch, botoconfig_name)
+
+        upload_funcs = {
+            "google-cloud-storage": upload_gs,
+            "rsync": upload_rsync
+        }
```

Comment:
I would define this at the top level, near where the functions it references are defined.